ext/neural_snow/NeuralSnow.jl

```diff
@@ -0,0 +1,240 @@
+ClimaLand.Snow.density_aux_names(m::NeuralDepthModel) = ()
+
+"""
+    snow_density(density::NeuralDepthModel, SWE::FT, z::FT, parameters::SnowParameters{FT}) where {FT}
+Returns the snow density given the current model state and the `NeuralDepthModel`
+density paramterization.
+"""
+function snow_density(
+    density::NeuralDepthModel{FT},
+    SWE::FT,
+    z::FT,
+    parameters::SnowParameters{FT},
+)::FT where {FT}
+    ρ_l = FT(LP.ρ_cloud_liq(parameters.earth_param_set))
+    if SWE <= eps(FT) #if there is no snowpack, aka avoid NaN
+        return FT(ρ_l)
+    end
+    ρ_new = SWE / z * ρ_l
+    return ρ_new
+end
+
+"""
+    update_density!(density::NeuralDepthModel, params::SnowParameters, Y, p)
+Updates the snow density given the current model state and the `NeuralDepthModel`
+density paramterization.
+"""
+function update_density!(
+    density::NeuralDepthModel,
+    params::SnowParameters,
+    Y,
+    p,
+)
+    p.snow.ρ_snow .= snow_density.(Ref(density), Y.snow.S, Y.snow.Z, params)
+end
+
+"""
+    eval_nn(vmodel, z::FT, swe::FT, P::FT, T::FT, R::FT, qrel::FT, u::FT)::FT where {FT}
+Helper function for evaluating the neural network in a pointwise manner over a `ClimaCore.Field`
+and returning the output in a broadcastable way.
+"""
+function eval_nn(
+    model,
+    z::FT,
+    swe::FT,
+    P::FT,
+    T::FT,
+    R::FT,
+    qrel::FT,
+    u::FT,
+)::FT where {FT}
+    input = FT.([z, swe, qrel, R, u, T, P])
+    return model(input)[1]
+end
+
+"""
+    dzdt(density::NeuralDepthModel, model::SnowModel{FT}, Y, p, t) where {FT}
+Returns the change in snow depth (rate) given the current model state and the `NeuralDepthModel`
+density paramterization, passing the approximate average of the forcings over the last 24 hours instead of
+the instantaneous value.
+"""
+function dzdt(density::NeuralDepthModel, Y)
+    return eval_nn.(
+        Ref(density.z_model),
+        Y.snow.Z,
+        Y.snow.S,
+        Y.snow.P_avg,
+        Y.snow.T_avg,
+        Y.snow.R_avg,
+        Y.snow.Qrel_avg,
+        Y.snow.u_avg,
+    )
+end
+
+"""
+    clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT
+A helper function which clips the tendency of Z such that
+its behavior is consistent with that of S.
+"""
+function clip_dZdt(S::FT, Z::FT, dSdt::FT, dZdt::FT, Δt::FT)::FT where {FT}
+    if (S + dSdt * Δt) <= eps(FT)
+        return -Z / Δt
+    elseif (Z + dZdt * Δt) < (S + dSdt * Δt)
+        #do we need to do something about setting q_l = 1 here?
+        return (S - Z) / Δt + dSdt
+    else
+        return dZdt
+    end
+end
+
+"""
+    update_density_prog!(density::NeuralDepthModel, model::SnowModel, Y, p)
+Updates all prognostic variables associated with density/depth given the current model state and the `NeuralDepthModel`
+density paramterization.
+"""
+function update_density_prog!(
+    density::NeuralDepthModel,
+    model::SnowModel,
+    dY,
+    Y,
+    p,
+)
+    #Get Inputs:
+    #is this too many allocations?
```

Comment:
yes, anything with a dot (broadcast) over a field needs to be in an expression that involves x .= ...

so that there are no allocations.

You can do this like:
```
    @. dY.snow.P_avg = β * (abs(p.drivers.P_snow) .- Y.snow.P_avg)
    @. dY.snow.T_avg = β * (p.drivers.T - model.parameters.earth_param_set.T_freeze - Y.snow.T_avg)
    @. dY.snow.R_avg = β * (p.drivers.SW_d - Y.snow.R_avg) # SW_d is positive by definition
    @. dY.snow.Qrel_avg = β * (Thermodynamics.relative_humidity(
            Ref(model.boundary_conditions.atmos.thermo_params),
            p.drivers.thermal_state,
        ) - Y.snow.Qrel_avg)
    @. dY.snow.u_avg = β * (wind_speed - Y.snow.u_avg)
```

the way it is written, it is not possible to do this for dY.snow.Z. But you could add in an aux var
```
 p.snow.predicted_dzdt .=  dzdt(density, Y)
```
which is updated here (if you like) and then compute

```
@. dY.snow.Z = clip_dZdt(swe, z, dswedt, predicted_dzdt, dt)
```